### PR TITLE
INF-210 Add Slack alerts to CircleCI

### PR DESCRIPTION
**Pull request by joaquincasares**
### Description

Copy the audius-protocol implementation of Slack alerts into audius-client.

### Dragons

*Is there anything the reviewer should be on the lookout for? Are there any dangerous changes?*

See comments.

### How Has This Been Tested?

*Please describe the tests that you ran to verify your changes. Provide repro instructions & any configuration.*

### How will this change be monitored?

*For features that are critical or could fail silently please describe the monitoring/alerting being added.*

### Feature Flags ###

*Are all new features properly feature flagged? Describe added feature flags.*

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -10,18 +10,163 @@ commands:
...
+        default: "Job started. :crossed_fingers:"
+      slack_mentions_user:
+        description: Used for CircleCI @mentions
+        type: string
+        default: ""
+    steps:
+      - slack/notify:
+          channel: << parameters.channel >>
+          branch_pattern: << parameters.branch_pattern >>
+          custom: |
```
Comment (joaquincasares) [inline .circleci/config.yml:39]: We remove the OVERLY large CircleCI logo:
* https://github.com/CircleCI-Public/slack-orb/blob/master/src/message_templates/basic_success_1.json#L43

We also consolidate the Mentions field into an existing section to minimize vertical height noise.

File: .circleci/config.yml [hunk 2]
```diff
@@ -10,18 +10,163 @@ commands:
...
+                }
+              ]
+            }
+          event: << parameters.event >>
+  slack-fail:
+    parameters:
+      event:
+        description: when this notification will fire. options are fail, pass, always.
+        type: string
+        default: fail
```
Comment (joaquincasares) [inline .circleci/config.yml:101]: Default to `fail` since that's usually when we'll use this.

File: .circleci/config.yml [hunk 3]
```diff
@@ -10,18 +10,163 @@ commands:
...
+        type: string
+        default: always
+      branch_pattern:
+        description: branch_pattern arg for slack/notify
+        type: string
+        default: .+
+      channel:
+        description: Slack channel to send message to
+        type: string
+        default: $SLACK_DEFAULT_CHANNEL
```
Comment (joaquincasares) [inline .circleci/config.yml:26]: This is set via the `slack-secrets` context as a required envar: 

* https://github.com/CircleCI-Public/slack-orb/wiki/Setup#4-create-a-context-on-circleci

File: .circleci/config.yml [hunk 4]
```diff
@@ -10,18 +10,163 @@ commands:
...
+      text:
+        description: Markdown formatted Slack message
+        type: string
+        default: "Job started. :crossed_fingers:"
```
Comment (joaquincasares) [inline .circleci/config.yml:30]: This is self-deprecating humor. Feel free to make it more positive. 🤓 

I'll follow suit. :D

File: .circleci/config.yml [hunk 5]
```diff
@@ -10,18 +10,163 @@ commands:
       - run:
           name: Combine package-lock.json files to single file
           command: ls -d -- packages/*/*-lock.json | awk -F packages '{printf "\"packages%s\" ", $2}' | xargs cat > << parameters.filename >>
+  slack-basic:
```
Comment (joaquincasares) [inline .circleci/config.yml:13]: Probably better to go in it's own commands.yml, but I'll keep an eye on how you do this so I can do the same in audius-protocol. 

Side note: Perhaps we can have a call soon so you can walk me through how you organized your ymls? I noticed `generate-config` today. Do you think it's worth separating out our config.yml within `audius-protocol` today?

File: .circleci/config.yml [hunk 6]
```diff
@@ -10,18 +10,163 @@ commands:
...
+          branch_pattern: << parameters.branch_pattern >>
+          channel: << parameters.channel >>
+          text: << parameters.text >>
+          slack_mentions_user: << parameters.slack_mentions_user >>
+  slack-success:
+    parameters:
+      event:
+        description: when this notification will fire. options are fail, pass, always.
+        type: string
+        default: pass
```
Comment (joaquincasares) [inline .circleci/config.yml:130]: Default to `pass`.

File: .circleci/config.yml [hunk 7]
```diff
@@ -87,6 +232,10 @@ jobs:
       - add_ssh_keys:
           fingerprints:
             - "a4:99:2e:fa:80:31:dc:23:25:7b:a5:70:3f:08:5f:fa"
+      - slack-basic:
+          event: always
+          branch_pattern: master
+          slack_mentions_user: << pipeline.parameters.slack_mentions_user >>
```
Comment (joaquincasares) [inline .circleci/config.yml:238]: Within `audius-protocol` I'm passing this around as a job parameter as well.

This seemed easier and perhaps a better style to call `pipeline.parameters` directly?

File: .circleci/config.yml [hunk 8]
```diff
@@ -105,6 +254,12 @@ jobs:
             VERSION=$(jq -r .version ./lerna.json)
             git checkout -b release-v${VERSION}
             git push -u origin release-v${VERSION}
+      - slack-fail:
+          branch_pattern: master
+          slack_mentions_user: << pipeline.parameters.slack_mentions_user >>
+      - slack-success:
+          branch_pattern: master
+          slack_mentions_user: << pipeline.parameters.slack_mentions_user >>
```
Comment (joaquincasares) [inline .circleci/config.yml:262]: If the previous `run` command fails, we'll trigger only `slack-fail`, else we'll only trigger `slack-success`.

File: .circleci/config.yml [hunk 9]
```diff
@@ -161,3 +316,5 @@ workflows:
         - equal: ["generate-release", << pipeline.schedule.name >>]
     jobs:
       - generate-release
+          context:
+            - slack-secrets
```
Comment (joaquincasares) [inline .circleci/config.yml:320]: We're using the same context as audius-protocol which defaults the channel to #eng-deploy-announcements using the CircleCI user I created.

**Comment by audius-infra** [comment]
Preview this change https://demo.audius.co/jc-inf-210.client

File: .circleci/config.yml (hunk 9 above)
Comment (joaquincasares) [reply .circleci/config.yml:320]: Not sure why this line is causing build errors when it's similar to audius-protocol:

* https://github.com/AudiusProject/audius-protocol/blob/89e0db2520df13187b809488c4b3aa723a2a7361/.circleci/config.yml#L1017

Comment (sliptype) [reply .circleci/config.yml:320]: I think the context needs to be at the workflow level instead of the job level

File: .circleci/config.yml [hunk 10]
```diff
@@ -105,6 +254,12 @@ jobs:
             VERSION=$(jq -r .version ./lerna.json)
             git checkout -b release-v${VERSION}
             git push -u origin release-v${VERSION}
+      - slack-fail:
+          branch_pattern: master
```
Comment (sliptype) [inline .circleci/config.yml:258]:
```suggestion
          branch_pattern: main
```

File: .circleci/config.yml (hunk 7 above)
Comment (sliptype) [reply .circleci/config.yml:238]: Yeah, seems cleaner!

File: .circleci/config.yml (hunk 5 above)
Comment (sliptype) [reply .circleci/config.yml:13]: Yup I think we can put this in the `.circleci/commands/commands.yml` file. Let's chat today about how this is structured, at a high level it's using circleci's dynamic configuration feature: https://circleci.com/docs/dynamic-config

I would say yes it makes sense to break up the `audius-protocol` config as well. It's pretty big already as one file

**Review by sliptype — APPROVED** [review]

**Pull request closed without merging**